File: problems/2300_successful_pairs_of_spells_and_potions/2300_successful_pairs_of_spells_and_potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        n, m = len(spells), len(potions)
        res = []
        potions.sort()
        for i, spell in enumerate(spells):
            need = (success + spell - 1) // spell
            idx = bisect.bisect_left(potions, need)
            res.append(m - idx)
        return res

File: problems/2300_successful_pairs_of_spells_and_potions/2300_successful_pairs_of_spells_and_potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length, m = potions.length;
        int[] res = new int[n];
        Arrays.sort(potions);
        for (int i = 0; i < n; i++) {
            // ceil integer
            int spell = spells[i];
            long need = (success + spell - 1) / spell;
            int idx = binarySearch(potions, need);
            res[i] = m - idx;
        }
        return res;
    }

    private int binarySearch(int[] arr, long target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if ((long) arr[mid] >= target) { // find first >= target
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }
}